src/registrar/models/domain.py

```diff
@@ -1648,6 +1667,105 @@ def _get_dnssec_data(self, response_extensions):
                 dnssec_data = extension
         return dnssec_data
 
+    def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
+        """Capture and cache old hosts and contacts from cache if they don't exist in cleaned"""
+        old_cache_hosts = self._cache.get("hosts")
+        old_cache_contacts = self._cache.get("contacts")
+
+        if fetch_contacts:
+            cleaned["contacts"] = self._get_contacts(cleaned.get("_contacts", []))
+            if old_cache_hosts is not None:
+                logger.debug("resetting cleaned['hosts'] to old_cache_hosts")
+                cleaned["hosts"] = old_cache_hosts
+
+        if fetch_hosts:
+            cleaned["hosts"] = self._get_hosts(cleaned.get("_hosts", []))
+            if old_cache_contacts is not None:
+                cleaned["contacts"] = old_cache_contacts
+
+    def _update_hosts_and_ips_in_db(self, cleaned, fetch_hosts):
+        """Update hosts and host_ips in database if retrieved from registry.
+
+        Parameters:
+            self: the domain to be updated with hosts and ips from cleaned
+            cleaned: dict containing hosts.  Hosts are provided as a list of dicts, e.g.
+                [{"name": "ns1.example.com",}, {"name": "ns1.example.gov"}, "addrs": ["0.0.0.0"])]
+            fetch_hosts: boolean indicating whether or not fetch_hosts was called
+        """
+        if fetch_hosts:
+            cleaned_hosts = cleaned["hosts"]
+            # Get all existing hosts from the database for this domain
+            existing_hosts_in_db = Host.objects.filter(domain=self)
+            # Identify hosts to delete
+            cleaned_host_names = set(cleaned_host["name"] for cleaned_host in cleaned_hosts)
+            hosts_to_delete_from_db = [
+                existing_host for existing_host in existing_hosts_in_db if existing_host.name not in cleaned_host_names
+            ]
+            # Delete hosts and their associated HostIP instances
+            for host_to_delete in hosts_to_delete_from_db:
+                # Delete associated HostIP instances
+                HostIP.objects.filter(host=host_to_delete).delete()
+                # Delete the host itself
+                host_to_delete.delete()
```

Comment:
```suggestion
            for host_to_delete in hosts_to_delete_from_db:
                # Delete associated HostIP instances
                HostIP.objects.filter(host=host_to_delete).delete()
            Host.objects.filter(id__in=hosts_to_delete_from_db).delete()
```
(Optional - if possible) I am not immediately sure if it will let you (unsure if it extrapolates id like this), but it may be more efficient to do something like above, as it reduces the total number of db calls